ci/image/Dockerfile

```diff
@@ -0,0 +1,170 @@
+    && sed -i "s/^SLAPD_SERVICES.*/SLAPD_SERVICES=\"ldap\:\/\/\/ ldapi\:\/\/\/ ldaps\:\/\/\/\"/g" /etc/default/slapd \
+    && echo "#!/usr/bin/env bash"                                               >> /bin/start-slapd \
+    && echo "set -eu -o pipefail"                                               >> /bin/start-slapd \
+    && echo "echo '# docker build will not persist /etc/hosts'  >> /etc/hosts"  >> /bin/start-slapd \
+    && echo "echo '# ------------- UAA DNS ------------- #'     >> /etc/hosts"  >> /bin/start-slapd \
+    && echo "echo '127.0.0.1 oidcloginit.localhost'             >> /etc/hosts"  >> /bin/start-slapd \
+    && echo "echo '127.0.0.1 testzone1.localhost'               >> /etc/hosts"  >> /bin/start-slapd \
+    && echo "echo '127.0.0.1 testzone2.localhost'               >> /etc/hosts"  >> /bin/start-slapd \
+    && echo "echo '127.0.0.1 testzone3.localhost'               >> /etc/hosts"  >> /bin/start-slapd \
+    && echo "echo '127.0.0.1 testzone4.localhost'               >> /etc/hosts"  >> /bin/start-slapd \
+    && echo "echo '127.0.0.1 testzoneinactive.localhost'        >> /etc/hosts"  >> /bin/start-slapd \
+    && echo "echo '127.0.0.1 testzonedoesnotexist.localhost'    >> /etc/hosts"  >> /bin/start-slapd \
+    && echo "echo '# ------------- UAA DNS ------------- #'     >> /etc/hosts"  >> /bin/start-slapd \
+    && echo "pgrep slapd || service slapd start"                                >> /bin/start-slapd \
+    && chmod 0755 /bin/start-slapd
+#/LDAP
+
+# PROJECT_SETUP
+ENV DB_NAME=uaa
+ENV DB_USER=root
+ENV DB_PASS=changeme
+ENV NUM_DBS=24
+
+RUN /bin/start-mysql \
+    && echo "[mysql]"               >> "${HOME}/.my.cnf" \
+    && echo "password=${DB_PASS}"   >> "${HOME}/.my.cnf" \
+    && mysql -e "ALTER USER '${DB_USER}'@'localhost' IDENTIFIED WITH mysql_native_password BY '${DB_PASS}';" \
+    # -----------^ change root from `auth_socket` to `mysql_native_password` so that `-h 127.0.0.1` works
+    && mysql -e "CREATE DATABASE ${DB_NAME} DEFAULT CHARACTER SET utf8 DEFAULT COLLATE utf8_general_ci;" \
+    && for i in $(seq 1 ${NUM_DBS}); do mysql -e "CREATE DATABASE ${DB_NAME}_${i} DEFAULT CHARACTER SET utf8 DEFAULT COLLATE utf8_general_ci;" ; done
+
+RUN /bin/start-postgresql \
+    && sudo -u postgres psql -c "CREATE USER ${DB_USER} WITH SUPERUSER PASSWORD '${DB_PASS}';" \
+    && sudo -u postgres createdb "${DB_USER}" \
+    && sudo -u postgres createdb "${DB_NAME}" \
+    && for i in $(seq 1 ${NUM_DBS}); do sudo -u postgres createdb "${DB_NAME}_${i}" ; done
+
+COPY ldap_db_init.ldif copy_of-uaa_src_main_resources_ldap_init.ldif /ldap/
+RUN /bin/start-slapd \
+    && ldapadd -Y EXTERNAL -H ldapi:/// -f /ldap/ldap_db_init.ldif \
+    && ldapadd -x -D 'cn=admin,dc=test,dc=com' -w password -f /ldap/copy_of-uaa_src_main_resources_ldap_init.ldif \
+    && rm -rf /ldap/
+#/PROJECT_SETUP
+
+# CLEANUP
+RUN apt-get autoremove -yqq \
+   && apt-get clean \
+   && rm -rf /var/lib/apt/lists/* /tmp/* /var/tmp/*
+#/CLEANUP
```

Comment:
Not sure why clean up would be necessary this late in the DockerFile, the data being removed will have been cached in earlier layers...

Reply:
upon reflection this makes sense. I think I lifted this pattern from elsewhere. given the per-layer caching the image size reduction benefit is only helpful if this cleanup happens in every layer which issues an `apt-get` command.
:-/

Reply:
yeah I know :/

Also never tried [this but apparently bazel can build images without all the layers](https://github.com/bazelbuild/rules_docker) but not going there for this.